mrp_procurements_plan/models/procurement.py

```diff
@@ -64,5 +53,99 @@ def run(self, autocommit=False):
+        if self.product_id:
+            raise exceptions.Warning(_('Error!: You can not import'
+                                       ' procurements, there is defined an'
+                                       ' final product'))
+        super(ProcurementPlan, self).action_import()
+
+    @api.multi
+    def button_generate_procurements(self):
+        for plan in self:
+            if not plan.product_id:
+                raise exceptions.Warning(_('Error!: You must enter the final'
+                                           ' product'))
+            if not plan.qty_to_produce:
+                raise exceptions.Warning(_('Error!: You must enter the'
+                                           ' quantity to produce'))
+            if not plan.mrp_bom_id:
+                raise exceptions.Warning(_('Error!: No BoM found'))
+            if len(plan.procurement_ids) > 0 and self.state != 'cancel':
+                raise exceptions.Warning(_('Error!: Already generated'
+                                           ' procurements orders'))
+            plan._deploy_bom_details(plan.mrp_bom_id)
+        return True
+
+    def _deploy_bom_details(self, bom):
+        for line in bom.bom_line_ids:
+            self._create_procurement_from_bom_line(1, line.product_id,
+                                                   line.product_qty)
+            if line.child_line_ids:
+                self._calculate_bom_line_details(line.child_line_ids, 1)
+        return True
+
+    def _calculate_bom_line_details(self, child_line_ids, level):
+        level += 1
+        for line in child_line_ids:
+            self._create_procurement_from_bom_line(level, line.product_id,
+                                                   line.product_qty)
+            if line.child_line_ids:
+                self._calculate_bom_line_details(line.child_line_ids, level)
+        return True
+
+    def _create_procurement_from_bom_line(self, level, product, product_qty):
+        procurement_obj = self.env['procurement.order']
+        qty = ((product_qty * self.qty_to_produce) /
+               self.mrp_bom_id.product_qty)
+        vals = {'name': 'Generated from plan',
+                'level': level,
+                'product_id': product.id,
+                'plan': self.id,
+                'main_project_id': self.project_id.id,
+                'product_qty': qty
+                }
+        val = procurement_obj.onchange_product_id(product.id)
+        vals.update(val['value'])
+        procurement_obj.create(vals)
+        return True
+
+    @api.one
+    @api.onchange('product_id')
+    def onchange_product_id(self):
+        bom_obj = self.env['mrp.bom']
+        if len(self.procurement_ids) > 0 and self.state != 'cancel':
+            raise exceptions.Warning(_('Error!: You can not change Product'))
+        self.mrp_bom_id = False
+        if self.product_id:
+            cond = [('product_tmpl_id', '=',
+                     self.product_id.product_tmpl_id.id),
```

Comment:
No son 3 tipos, son solo 2...
O tiene informado product.product o no lo tiene. Si product.product está
informado, también debería estarlo template,ya que sería el template del
product. Otros casos serían incorrectos.

Si lo tiene, el bom del template nos da igual, hay que coger el del
product.product. Solo hay que mirar el de template si product está vacío.

2015-05-04 9:27 GMT+02:00 alfredoavanzosc <notifications@github.com>:

> In mrp_procurements_plan/models/procurement.py
> <https://github.com/odoomrp/odoomrp-wip/pull/799#discussion_r29569409>:
>
> > +                }
> > +        val = procurement_obj.onchange_product_id(product.id)
> > +        vals.update(val['value'])
> > +        procurement_obj.create(vals)
> > +        return True
> > +
> > +    @api.one
> > +    @api.onchange('product_id')
> > +    def onchange_product_id(self):
> > +        bom_obj = self.env['mrp.bom']
> > +        if len(self.procurement_ids) > 0 and self.state != 'cancel':
> > +            raise exceptions.Warning(_('Error!: You can not change Product'))
> > +        self.mrp_bom_id = False
> > +        if self.product_id:
> > +            cond = [('product_tmpl_id', '=',
> > +                     self.product_id.product_tmpl_id.id),
>
> Si pongo la condición como tu dices, y me encuentro que existen los 3
> tipos, ¿Con cuál se quedaría?, yo primero busco con product.template y
> product.product, luego con product.product, y luego con product. template.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/odoomrp/odoomrp-wip/pull/799/files#r29569409>.
>



-- 
CEO Avanzosc, S.L <http://www.avanzosc.es> : Office phone / Tfono oficina:
(+34) 943 02 69 02
Ana Juaristi Olalde <http://www.anajuaristi.com/>: Personal phone: 677 93
42 59. User/usuario skype: Avanzosc
www.openerpsite.com


*El contenido de esta comunicación y de toda su documentación anexa es
confidencial y se dirige exclusivamente a su destinatario. El uso no
autorizado de esta información está prohibido por la legislación vigente.
Si usted no es el destinatario le rogamos nos lo indique, no comunique su
contenido a terceros y proceda a su destrucción. Disculpe las molestias que
le haya ocasionado la recepción indebida de este e-mail. Sus datos figuran
en un fichero cuyo titular es Avanzosc, S.L., a quien usted puede dirigirse
para ejercer sus derechos de acceso, rectificación, cancelación y oposición
en Klara Donea 13, 20720, Azkoitia (Gipuzkoa), Tef. 943 02 69 02 -
**administracion@avanzosc.
<soporte@avanzosc.es>es*

*Komunikazio honen edukia eta dokumentazio erantsia konfidentziala da eta
hartzaileak bakarrik jaso beharko luke. Indarrean dagoen legeriak debekatu
egiten du bertan eskainitako informazioa baimenik gabe erabiltzea.
Komunikazioa zuri iritsi bazaizu, baina zu ez bazara hartzailea, mesedez,
guri jakinarazi, eta jasotako informazioa ez inori jakinarazi eta suntsitu.
Barkatu okerreko email hau jasotzeak eragindako eragozpenak. Zure datuak
Avanzosc, S.L. enpresaren fitxategietan sartuta daude. Zure datuak atzitzea
eska dezakezu, bai eta, datuak zuzentzea, ezereztea eta tratamenduari aurka
egitea ere. Horretarako, enpresara jo dezakezu, helbide honetan: **Klara
Donea 13** 20720, Azkoitia (Gipuzkoa), telefonoa: 943 02 69 02 - *
*administracion@avanzosc. <soporte@avanzosc.es>es* *This message and all
documents attached to it are confidential and intended only for the person
or entity to which it is addressed. Any use of this information by
unauthorised persons is prohibited under current legislation. If you
received this message by error, please advise us, destroy it and refrain
from communicating its contents to third parties. We apologise for any
inconvenience receiving this email improperly may cause to you. Your
personal data are included in a file owned by Avanzosc, S.L. If you want to
exercise your rights of access, correction, erasure and objection you can
contact the Controller at **Klara Donea 13** 20720, Azkoitia (Gipuzkoa), T:
943 02 69 02 – administracion@avanzosc. <soporte@avanzosc.es>es*